src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,350 @@
+                    if (createPayloads.Length == 1)
+                    {
+                        MethodSymbol createPayload = createPayloads[0] as MethodSymbol;
+                        // ToDo: Add checks for parameter types.
+                        if (createPayload != null && createPayload.IsStatic && createPayload.ParameterCount == 4 && createPayload.Parameters[0].Name == "mvid" && createPayload.Parameters[1].Name == "methodToken" && createPayload.Parameters[2].Name == "payload" && createPayload.Parameters[3].Name == "payloadLength")
+                        {
+                            _createPayload = createPayload;
+                        }
+                    }
+                }
+            }
+
+            return _createPayload;
+        }
+
+        private static MethodSymbol GetFlushPayload(CSharpCompilation compilation)
+        {
+            if (_flushPayload == null)
+            {
+                NamedTypeSymbol instrumentationType = compilation.GetTypeByMetadataName("Microsoft.CodeAnalysis.Runtime.Instrumentation");
+                if (instrumentationType != null)
+                {
+                    ImmutableArray<Symbol> flushPayloads = instrumentationType.GetMembers("FlushPayload");
+                    if (flushPayloads.Length == 1)
+                    {
+                        MethodSymbol flushPayload = flushPayloads[0] as MethodSymbol;
+                        if (flushPayload != null && flushPayload.IsStatic && flushPayload.ParameterCount == 0)
+                        {
+                            _flushPayload = flushPayload;
+                        }
+                    }
+                }
+            }
+
+            return _flushPayload;
+        }
+
+        private static bool IsTestMethod(MethodSymbol method)
+        {
+            // ToDo: Make this real.
+            return method.Name.StartsWith("Test");
+        }
+    }
+
+    internal sealed class InstrumentationInjectionWalker : BoundTreeRewriterWithStackGuard
+    {
+        private readonly MethodSymbol _method;
+        private readonly ArrayBuilder<SourceSpan> _spansBuilder;
+        private readonly FieldSymbol _payload;
+        private readonly TypeCompilationState _compilationState;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+
+        public InstrumentationInjectionWalker(MethodSymbol method, ArrayBuilder<SourceSpan> spansBuilder, FieldSymbol payload, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider)
+        {
+            _method = method;
+            _spansBuilder = spansBuilder;
+            _payload = payload;
+            _compilationState = compilationState;
+            _diagnostics = diagnostics;
+            _debugDocumentProvider = debugDocumentProvider;
+        }
+
+        public override BoundNode VisitBadStatement(BoundBadStatement node)
```

Comment:
I feel like I'm probably missing something obvious, but why not override `Visit` rather than each individual `VisitXStatement`?  Wouldn't that make the type more resilient against the addition of new kinds of statements?